### PR TITLE
Add TypeScript definition

### LINKED ISSUE #74: Add TypeScript type definition
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/pify/issues/74)

Would be nice to have TypeScript type definitions bundled in this package. The types should be strict and should be generic so it preserves the original function type.

The type definition should follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide), and should have full docs and type tests.

Some examples:
- https://github.com/sindresorhus/semver-regex/commit/039944bb5b3876913171b895488874b5c200e458
- https://github.com/sindresorhus/global-dirs/commit/0aae4d64887f77af94c4fb7ca12aa9f93c0a8291
- https://github.com/sindresorhus/md5-hex/commit/bbbfeff840f26973fbffc16d4863ae6d120079f3

## Note that this is not an easy task and requires you to have advanced TypeScript experience. There have been multiple attempts and a lot of wasted time on my part in the past. If you intend to work on this, please look at the previous attempts and the feedback given there.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $100.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)
- [omgimalexis omgimalexis](https://issuehunt.io/u/omgimalexis) ($20.00)

### Submitted pull Requests
- [#75 [WIP] Add TypeScript definition](https://issuehunt.io/r/sindresorhus/pify/pull/75)
- [#76 Add index.d.ts and tests](https://issuehunt.io/r/sindresorhus/pify/pull/76)
- [#85 Add TypeScript definition](https://issuehunt.io/r/sindresorhus/pify/pull/85)
- [#86 Add TypeScript definitions](https://issuehunt.io/r/sindresorhus/pify/pull/86)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/pify/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by chinesedfan**
Closes #74 and takes over #75.

If I missed something, glad to receive better inspirations from you guys.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#74: Add TypeScript type definition](https://issuehunt.io/repos/41585934/issues/74)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by chinesedfan** [comment]
With the powerful `ts-toolbelt`, it seems to have stricter types. But if you don't want to add an extra dependency, we can copy those utilities out.

Now the last thing is whether we can write stricter types when the input is an object. I will keep trying.

**Comment by chinesedfan** [comment]
@sindresorhus Ready to review now. Those types look complex, but really interesting.

**Comment by satyarohith** [comment]
Hi @chinesedfan, the title of the PR should be `Add TypeScript definition`. You can read the style guide [here](https://github.com/sindresorhus/typescript-definition-style-guide).

**Comment by sindresorhus** [comment]
Can you follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide)? Please read it carefully.

Don't forget docs comments (see: https://github.com/sindresorhus/pify/pull/75/files)

And also a lot more type tests (see: https://github.com/sindresorhus/pify/pull/75/files#r327017529)

**Comment by chinesedfan** [comment]
@sindresorhus Updated with fixing styles and adding docs. But I have a question about how to write more tests. 

As far as I know, `tsd` only checks types, without abilities to execute the script really. So how can we write tests that validating the resulted values like https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/pify/pify-tests.ts?

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+        if (promiseFn()) {
+            console.log(await promiseFn.method('hi'));
+        }
+    })();
+	```
+	*/
+    excludeMain?: boolean,
+
+    /**
```
Comment (sindresorhus) [inline index.d.ts]: Use tab-indentation

**Comment by sindresorhus** [comment]
> As far as I know, tsd only checks types, without abilities to execute the script really. So how can we write tests that validating the returned values like DefinitelyTyped/DefinitelyTyped:types/pify/pify-tests.ts@master ?

TypeScript doesn't execute either. TypeScript is purely a compile-step.

File: index.test-d.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+expectType<Promise<Buffer>>(pify(fs).readFile('a.txt'))
+expectType<Promise<never>>(pify(() => {})())
+expectType<Promise<[{ statusCode: number }?, string?]>>(pify(request, { multiArgs: true })('https://www.google.com'))
+expectType<Promise<[{ statusCode: number }?, string?]>>(pify(request2, { multiArgs: true, errorFirst: false })('https://www.google.com'))
+expectType<Promise<Buffer>>(pify(fs, { include: ['read'] }).readFile('a.txt'))
+expectAssignable<typeof fs.readFile>(pify(fs, { exclude: ['read'] }).readFile)
+expectAssignable<typeof request>(pify(request, { excludeMain: true }))
+expectType<Promise<boolean>>(pify(fs.exists, { errorFirst: false })('a.txt'))
+expectType<Promise<Buffer>>(pify(fs.readFile, { promiseModule: Promise })('a.txt'))
```
Comment (sindresorhus) [inline index.test-d.ts]: There should also be lots of tests that checks the `pify` return value:

```ts
expectType<InsertTypeHere>(pify(fs.readFile, { promiseModule: Promise }))
```

**Comment by sindresorhus** [comment]
> Can you follow this styleguide? Please read it carefully.

Not followed.

File: index.d.ts (hunk 1 above)
Comment (chinesedfan) [reply index.d.ts]: Really stupid. I failed at the first and the most obvious style rule.

File: index.test-d.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+expectAssignable<(path: string) => Promise<boolean>>(pify(fs.exists, { errorFirst: false }))
+expectType<Promise<boolean>>(pify(fs.exists, { errorFirst: false })('a.txt'))
+
+expectAssignable<(path: string) => Promise<Buffer | undefined>>(pify(fs.readFile, { promiseModule: Promise }))
+expectType<Promise<Buffer | undefined>>(pify(fs.readFile, { promiseModule: Promise })('a.txt'))
+
+declare function generic<A, B>(arg1: A, arg2: B, callback: (error?: Error, result?: A[]) => void): void
+expectAssignable<(arg1: number, arg2: string) => Promise<number[] | undefined>>(pify<(arg1: number, arg2: string, callback: (error?: Error, result?: number[]) => void) => void>(generic))
+expectType<Promise<number[] | undefined>>(pify<(arg1: number, arg2: string, callback: (error?: Error, result?: number[]) => void) => void>(generic)(123, 'abc'))
```
Comment (chinesedfan) [inline index.test-d.ts]: I didn't find a way to make pify returns a generic function again. Users have to specify types of `PifyInput` explicitly. 😢

Comment (sindresorhus) [reply index.test-d.ts:48]: I did warn people in the original issue:

> **Note that this is not an easy task and requires you to have advanced TypeScript experience.**

😛

**Comment by sindresorhus** [comment]
You might find some inspiration on how to solve this in https://github.com/sindresorhus/p-pipe/blob/master/index.d.ts

**Comment by chinesedfan** [comment]
> Including functions that accepts multiple generic arguments, and you should ensure that the generic arguments are preserved, even if there are many.

> declare function generic<A, B>(arg1: A, arg2: B, callback: (error?: Error, result?: A[]) => void): void

> declare function fnWithThreeArgs(arg1: number, arg2: string, arg3: boolean, callback: (error?: Error, result?: string[]) => void): void

@sindresorhus I think I misunderstood what you mean. `p-pipe` only handles situations like `fnWithThreeArgs`(in fact, this PR works well and I have updated tests), but neither of them handles situations like `generic`.

**Comment by sindresorhus** [comment]
> p-pipe only handles situations like fnWithThreeArgs

No, `p-pipe` handles multiple functions with generic arguments.

**Comment by sindresorhus** [comment]
Closing as this is not moving forward.

**Pull request closed without merging**